iup/all_objects.py

```diff
@@ -0,0 +1,366 @@
+        ), "Column 'elapsed', 'interval', 'estimate' and/or 'previous' is not numeric."
+
+
+class CumulativeUptakeData(UptakeData):
+    def validate(self):
+        assert set(self.columns).issuperset(
+            {"region", "date", "estimate"}
+        ), "At least one essential column is missing."
+        assert self["date"].dtype.is_temporal(), "Column 'date' is not temporal."
+        assert self["region"].dtype == pl.Utf8, "Column 'region' is not a string."
+        assert self["estimate"].dtype.is_numeric(), "Column 'estimate' is not numeric."
+        assert (
+            self["estimate"] >= 0.0
+        ).all(), "Cumulative uptake estimates cannot be negative."
+
+    def to_incident(self, rollout: dt.datetime) -> IncidentUptakeData:
+        self = (
+            self.with_columns(
+                season=pl.col("date").dt.year()
+                + pl.when(pl.col("date").dt.month() < 7).then(-1).otherwise(0),
+                estimate=pl.col("estimate")
+                .diff()
+                .fill_null(pl.col("estimate").first()),
+                elapsed=(pl.col("date") - rollout).cast(pl.Float64)
+                / (1000 * 60 * 60 * 24),
+            )
+            .with_columns(
+                interval=pl.col("elapsed").diff().fill_null(pl.col("elapsed").first()),
+                season=pl.col("season").cast(pl.Utf8),
+            )
+            .with_columns(estimate=pl.col("estimate") / pl.col("interval"))
+            .with_columns(previous=pl.col("estimate").shift(1))
+        )
+
+        if ((self["interval"] - self["interval"].mean()) / self["interval"].std())[
+            0
+        ] > 1:
+            self = self.slice(2, self.height - 1)
+        else:
+            self = self.slice(1, self.height - 1)
+
+        self = IncidentUptakeData(self)
+        self.validate()
+
+        return self
+
+
+class CumulativeUptakeProjection(UptakeData):
+    def validate(self):
+        assert set(self.columns).issuperset(
+            {"region", "date", "estimate"}
+        ), "At least one essential column is missing."
+        assert self["date"].dtype.is_temporal(), "Column 'date' is not temporal."
+        assert self["region"].dtype == pl.Utf8, "Column 'region' is not a string."
+        assert self["estimate"].dtype.is_numeric(), "Column 'estimate' is not numeric."
+
+
+def get_nis(
+    data_path, region_col, date_col, estimate_col, filters=None
+) -> CumulativeUptakeData:
+    out = (
+        (pl.read_csv(data_path))
+        .with_columns(estimate=pl.col(estimate_col).cast(pl.Float64, strict=False))
+        .drop_nulls(subset=["estimate"])
+    )
+
+    if filters is not None:
+        filter_expr = pl.lit(True)
+        for k, v in filters.items():
+            filter_expr &= pl.col(k).is_in(pl.lit(v))
+        out = out.filter(filter_expr)
+
+    out = (
+        out.with_columns(
+            region=pl.col(region_col),
+            date=pl.col(date_col)
+            .str.split(" ")
```

Comment:
Makes total sense, and I will change it here. But every NIS data set seems different. Sometimes the data column literally has entries like, "January 1, 2023 - January 8, 2023" in which case I need to split the start and end date apart. Is there a way to do this without a clunky regex?

Reply:
oh, I see

I think the long-term solution will be to build up a bank of NIS date parsing functions. For something like "January 1, 2023 - January 8, 2023", I think you want something that takes in a string (series/expr) and outputs a struct, with a start and end date. (And then we'd have other function on hand to deal with whatever other date formats are inevitably used in different files...)

I leave it up to you whether you want to (1) break out this date parsing as a separate function in this PR or (2) just open an issue about this and punt it

Reply:
In the spirit of modular PRs, I'll punt the bank of get_nis versions to an issue and go with your `.to_date()` solution for now, since it is the best option when possible.

Reply:
^^ link an issue?

Reply:
Issue #21. But I didn't link it, because if we close this PR unmerged and replace it with a brand new one, I don't want that linked issue to be closed automatically.